packages/flutter/lib/src/widgets/animated_raster.dart

```diff
@@ -0,0 +1,215 @@
+      ..animation = animation
+      ..delegate = delegate
+      ..devicePixelRatio = MediaQuery.maybeOf(context)?.devicePixelRatio ?? 1.0;
+  }
+}
+
+/// A render object that converts its children to a [ui.Image] while animating.
+///
+/// While the status of [animation] is [AnimationStatus.completed] or
+/// [AnimationStatus.dismissed], children are painted normally. When the status
+/// is [AnimationStatus.forward] or [AnimationStatus.reverse], this render object
+/// will create a GPU resident texture the first time that [AnimatedRasterDelegate.willPaint] returns true
+/// and provide it to [AnimatedRasterDelegate.paint].
+///
+/// See also:
+///  * [ZoomPageTransitionsBuilder], which uses this render object to implement the fade and
+///    stretch effect efficiently.
+class RenderAnimatedRaster extends RenderProxyBox {
+  /// Create a new [RenderAnimatedRaster].
+  RenderAnimatedRaster(this._animation, this._delegate, this._devicePixelRatio) {
+    animation.addListener(markNeedsPaint);
+    animation.addStatusListener(_updateStatus);
+    _updateStatus(animation.status);
+  }
+
+  AnimationStatus _status = AnimationStatus.completed;
+
+  void _updateStatus(AnimationStatus newStatus, [bool painting = false]) {
+    if (newStatus == _status) {
+      return;
+    }
+    _childImage?.dispose();
+    _childImage = null;
+    _status = newStatus;
+    assert(_status == animation.status);
+    if (!painting) {
+      markNeedsPaint();
+    }
+  }
+
+  /// The delegate that controls how the child raster is painted.
+  AnimatedRasterDelegate get delegate => _delegate;
+  AnimatedRasterDelegate _delegate;
+  set delegate(AnimatedRasterDelegate value) {
+    if (value == delegate) {
+      return;
+    }
+    _delegate = value;
+    markNeedsPaint();
+  }
+
+  /// The device pixel ratio used to create the child raster.
+  double get devicePixelRatio => _devicePixelRatio;
+  double _devicePixelRatio;
+  set devicePixelRatio(double value) {
+    if (value == devicePixelRatio) {
+      return;
+    }
+    _devicePixelRatio = value;
+    markNeedsPaint();
+  }
+
+
+  /// The animation used to drive this render object.
+  Animation<double> get animation => _animation;
+  Animation<double> _animation;
+  set animation(Animation<double> value) {
+    if (value == animation) {
+      return;
+    }
+    animation.removeListener(markNeedsPaint);
+    animation.removeStatusListener(_updateStatus);
+    _animation = value;
+    animation.addStatusListener(_updateStatus);
+    animation.addListener(markNeedsPaint);
+    _updateStatus(animation.status);
+    markNeedsPaint();
+  }
+
+  @override
+  void dispose() {
+    animation.removeListener(markNeedsPaint);
+    animation.removeStatusListener(_updateStatus);
+    _childImage?.dispose();
+    _childImage = null;
+    super.dispose();
+  }
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    animation.addListener(markNeedsPaint);
+    animation.addStatusListener(_updateStatus);
```

Comment:
Arn't these listeners now added twice? Once in the constructor and again on attach?

Reply:
Btw when programming Flutter, do we really have to spot such kind of bugs manually by eyes? Or can we develop some lints or verifiers for it... I am willing to contribute